single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -0,0 +1,875 @@
+            # Update the config server DB URI on the remote mongos
+            self.cluster_manager.update_config_server_db()
+
+    def open_ports(self) -> None:
+        """Open ports on the workload.
+
+        VM-only.
+        """
+        if self.substrate != Substrates.VM:
+            return
+        ports = [MongoPorts.MONGODB_PORT]
+        if self.state.is_role(MongoDBRoles.CONFIG_SERVER):
+            ports.append(MongoPorts.MONGOS_PORT)
+
+        try:
+            for port in ports:
+                self.workload.exec(["open-port", f"{port}/TCP"])
+        except WorkloadExecError as e:
+            logger.exception(f"Failed to open port: {e}")
+            raise
+
+    def _set_os_config(self) -> None:
+        """Sets sysctl config for mongodb."""
+        try:
+            self.sysctl_config.configure(OS_REQUIREMENTS)
+        except (sysctl.ApplyError, sysctl.ValidationError, sysctl.CommandError) as e:
+            # we allow events to continue in the case that we are not able to correctly configure
+            # sysctl config, since we can  still run the workload with wrong sysctl parameters
+            # even if it is not optimal.
+            logger.error(f"Error setting values on sysctl: {e.message}")
+            # containers share the kernel with the host system, and some sysctl parameters are
+            # set at kernel level.
+            logger.warning("sysctl params cannot be set. Is the machine running on a container?")
+
+    @property
+    def primary(self) -> str | None:
+        """Retrieves the primary unit with the primary replica."""
+        with MongoConnection(self.state.mongo_config) as connection:
+            try:
+                primary_ip = connection.primary()
+            except Exception as e:
+                logger.error(f"Unable to get primary: {e}")
+                return None
+
+        for unit in self.state.units:
+            if primary_ip == unit.internal_address:
+                return unit.name
+        return None
+
+    @override
+    def start_charm_services(self):
+        """Start the relevant services.
+
+        If we are running as config-server, we should start both mongod and mongos.
+        """
+        self.workload.start()
+        if self.state.is_role(MongoDBRoles.CONFIG_SERVER):
+            self.mongos_workload.start()
+
+    @override
+    def stop_charm_services(self):
+        """Stop the relevant services.
+
+        If we are running as config-server, we should stop both mongod and mongos.
+        """
+        if self.state.is_role(MongoDBRoles.CONFIG_SERVER):
+            self.mongos_workload.stop()
+        self.workload.stop()
+
+    @override
+    def restart_charm_services(self):
+        """Restarts the charm services with updated config.
+
+        If we are running as config-server, we should update both mongod and mongos environments.
+        """
+        self.stop_charm_services()
+        self.config_manager.set_environment()
```

Comment:
don't snaps provide the `restart` functionality?

I ask since I imagine a scenario where `set_environment` results in an exception and then the services never get restarted. And the the server is stuck in the down state

Reply:
That is an interesting question.
On one side, indeed we don't want the set_environment to fail and not having any disruption is important.
On the other side, if it fails and it was to setup security, we might not want to restart in an insecure configuration.

I followed the pattern from [the current charm](https://github.com/canonical/mongodb-operator/blob/b54377017eb21e4103793b26a85e44f4d93dd1d8/src/charm.py#L1392) for now.
Maybe what could be done would be to add some tenacity on the `set_environment` so that it doesn't fail?

Reply:
> On the other side, if it fails and it was to setup security, we might not want to restart in an insecure configuration.

Hmm I had no considered this.

I suppose if we have it like this and it hasn't caused any issues then its probably fine to keep as is. Thanks for your patience :)